test/cloudtest/test_k8s_node_recovery.py

```diff
@@ -0,0 +1,354 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+import subprocess
+from dataclasses import dataclass
+from textwrap import dedent
+
+import pytest
+
+from materialize import ui
+from materialize.cloudtest.app.materialize_application import MaterializeApplication
+
+
+@dataclass
+class ReplicaDefinition:
+    cluster_name: str
+    index: int
+    availability_zone: str
+
+    def get_name(self) -> str:
+        return f"{self.cluster_name}_r_{self.index}"
+
+
+@dataclass
+class ClusterDefinition:
+    name: str
+    replica_definitions: list[ReplicaDefinition]
+
+    def create_replica_definitions_sql(self) -> str:
+        replica_definitions = []
+        for replica in self.replica_definitions:
+            replica_definitions.append(
+                f"{replica.get_name()} (SIZE = '1', AVAILABILITY ZONE '{replica.availability_zone}')"
+            )
+
+        return ", ".join(replica_definitions)
+
+
+def default_compute_cluster() -> ClusterDefinition:
+    compute_cluster = ClusterDefinition("c_compute", [])
+    compute_cluster.replica_definitions.append(
+        ReplicaDefinition(compute_cluster.name, 1, "1")
+    )
+    return compute_cluster
+
+
+def default_storage_cluster() -> ClusterDefinition:
+    storage_cluster = ClusterDefinition("c_storage", [])
+    storage_cluster.replica_definitions.append(
+        ReplicaDefinition(storage_cluster.name, 1, "1")
+    )
+    return storage_cluster
+
+
+REQUIRED_PROGRESS = 5
+
+
+def populate(
+    mz: MaterializeApplication,
+    compute_cluster: ClusterDefinition,
+    storage_cluster: ClusterDefinition,
+) -> None:
+    all_clusters = [compute_cluster, storage_cluster]
+
+    drop_cluster_statements = [
+        f"> DROP CLUSTER IF EXISTS {cluster.name};" for cluster in all_clusters
+    ]
+    # string needs the same indentation as testdrive script below
+    drop_cluster_statement_sql = "\n            ".join(drop_cluster_statements)
+
+    create_cluster_statements = [
+        f"> CREATE CLUSTER {cluster.name} REPLICAS ({cluster.create_replica_definitions_sql()});"
+        for cluster in all_clusters
+    ]
+    # string needs the same indentation as testdrive script below
+    create_cluster_statement_sql = "\n            ".join(create_cluster_statements)
+
+    mz.testdrive.run(
+        input=dedent(
+            f"""
+            > DROP MATERIALIZED VIEW IF EXISTS mv;
+            > DROP SOURCE IF EXISTS source;
+
+            {drop_cluster_statement_sql}
```

Comment:
This with the many spaces in the string is a bit fragile. Consider doing it like this:

```
dedent(
f""""
> first testdrive fragment goes here
""") + "\n".join(drop_cluster_statements) + dedent
f"""
...
"""
)
````
In other words, exit the dedent temporarily and then start a new dedent after the statements you are generating.